firestore/query.go

```diff
@@ -250,6 +280,147 @@ func (q *Query) processCursorArg(name string, docSnapshotOrFieldValues []interfa
 
 func (q Query) query() *Query { return &q }
 
+func getSafeCursorValue(vProto *pb.Value, q Query) (interface{}, error) {
+	switch vProto.ValueType.(type) {
+	case *pb.Value_ReferenceValue:
+		refVal := vProto.GetReferenceValue()
+		if strings.HasPrefix(refVal, q.path) {
+			refVal = refVal[len(q.path)+1:]
+		}
+		return refVal, nil
+	default:
+		i, err := createFromProtoValue(vProto, q.c)
+		if err != nil {
+			q.err = err
+			return q, err
+		}
+		return i, nil
+	}
+}
+
+// FromProto creates a new Query object from a RunQueryRequest. This can be used
+// in combintation with ToProto to serialize Query objects.
+func (q Query) FromProto(pbQuery *pb.RunQueryRequest) (Query, error) {
```

Comment:
When will a user have a reference to this directly? This is never returned from the public api, correct?

Reply:
Users get this via ToProto. (line 390).

Reply:
I can look at adding docs. the purpose is to allow a user to fanout work to distributed process. `GetParititionedQueries`  returns a Slice of queries. To aid in serialization, these can be converted to RunQueryRequest which can be more easily serialized

Reply:
A small it of docstrings have been added to explain why you might use this.